mesh/mesh.go

```diff
@@ -0,0 +1,272 @@
+// limitations under the License.
+
+package mesh
+
+import (
+	"fmt"
+	"net"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/weaveworks/mesh"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type Mesh struct {
+	Router *mesh.Router
+	Collector *meshCollector
+	readyChan chan struct{}
+	logger log.Logger
+}
+
+// Create a new Mesh.
+func NewMesh(addr, hwaddr, nickname, pw string, logger log.Logger) (*Mesh, error) {
+	if addr == "" {
+		// TODO(iksaif): Add a DummyMesh type to make main.go simpler.
+		return nil, nil
+	}
+
+	host, portStr, err := net.SplitHostPort(addr)
+
+	if err != nil {
+		level.Error(logger).Log("msg", "Invalid mesh address", "address", addr, "err", err)
+		return nil, err
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		level.Error(logger).Log("msg", "Invalid mesh address", "address", addr, "err", err)
+		return nil, err
+	}
+
+	name, err := mesh.PeerNameFromString(hwaddr)
+	if err != nil {
+		level.Error(logger).Log("msg", "Invalid hardware address", "address", hwaddr, "err", err)
+		return nil, err
+	}
+
+	password := []byte(pw)
+	if len(password) == 0 {
+		// Emtpy password is used to disable secure communication. Using a nil
+		// password disables encryption in mesh.
+		password = nil
+	}
+
+	mrouter, err := mesh.NewRouter(mesh.Config{
+		Host:               host,
+		Port:               port,
+		ProtocolMinVersion: mesh.ProtocolMinVersion,
+		Password:           password,
+		ConnLimit:          64,
+		PeerDiscovery:      true,
+		TrustedSubnets:     []*net.IPNet{},
+	}, name, nickname, mesh.NullOverlay{}, printfLogger{logger})
+	if err != nil {
+		return nil, err
+	}
+
+	mesh := &Mesh{
+		Router: mrouter,
+		Collector: NewMeshCollector(mrouter),
+		readyChan: make(chan struct{}),
+		logger: logger,
+	}
+	return mesh, nil
+}
+
+// Start the mesh.
+func (m *Mesh) Start(peers []string) {
+	m.Router.Start()
+	m.Router.ConnectionMaker.InitiateConnections(peers, true)
+}
+
+// Settle waits until the mesh is ready (and sets the appropriate internal state when it is).
+// The idea is that we don't want to start "working" before we get a chance to know most of the alerts and/or silences.
+func (m *Mesh) Settle(interval, timeout time.Duration) {
+	// Inspired from https://github.com/apache/cassandra/blob/7a40abb6a5108688fb1b10c375bb751cbb782ea4/src/java/org/apache/cassandra/gms/Gossiper.java
+	// This is clearly not perfect or strictly correct but should prevent the alertmanager to send notification before it is obviously not ready.
+	// This is especially important for those that do not have persistent storage.
+	const NumOkayRequired = 3
+	level.Info(m.logger).Log("msg", "Waiting for gossip to settle...")
+	start := time.Now()
+	nPeers := 0
+	nOkay := 0
+	totalPolls := 0
+	for {
+		time.Sleep(interval)
+		elapsed := time.Since(start)
+		n := len(m.Router.Peers.Descriptions())
```

Comment:
I wonder if this can have a race where a deployment doesn't continue because the first peer has discovered where other peers are supposed to live, but can't discover them in mesh.

Reply:
There is a global timeout for Settle(), so we should be fine.